T/t8code/build_tarballs.jl

```diff
@@ -0,0 +1,112 @@
+# Note that this script can accept some limited command-line arguments, run
+# `julia build_tarballs.jl --help` to see a usage message.
+using BinaryBuilder, Pkg
+using Base.BinaryPlatforms
+const YGGDRASIL_DIR = "../.."
+include(joinpath(YGGDRASIL_DIR, "platforms", "mpi.jl"))
+
+name = "t8code"
+version = v"1.1.2"
+
+# Collection of sources required to complete build
+sources = [
+    ArchiveSource("https://github.com/DLR-AMR/t8code/releases/download/v$(version)/t8code_v$(version).tar.gz",
+                  "0bd4bee6694735d14fb4274275fb8c4bdeacdbd29b257220c308be63e98be8f7"),
+    DirectorySource("./bundled")
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir
+cd t8code/
+atomic_patch -p1 "${WORKSPACE}/srcdir/patches/mpi-constants.patch"
+
+# Set default preprocessor and linker flags
+# Note: This is *crucial* for Windows builds as otherwise the wrong libraries are picked up!
+export CPPFLAGS="-I${includedir}"
+export LDFLAGS="-L${libdir}"
+export CFLAGS="-O3"
+export CXXFLAGS="-O3"
+
+# Set necessary flags for FreeBSD
+if [[ "${target}" == *-freebsd* ]]; then
+  export LIBS="-lm"
+fi
+
+# Set necessary flags for Windows and non-Windodws systems
+FLAGS=()
+if [[ "${target}" == *-mingw* ]]; then
+  # Set linker flags only at build time (see https://docs.binarybuilder.org/v0.3/troubleshooting/#Windows)
+  FLAGS+=(LDFLAGS="$LDFLAGS -no-undefined")
+  # Configure does not find the correct Fortran compiler
+  export F77="f77"
+  # Link against ws2_32 to use the htonl function from winsock2.h
+  export LIBS="-lmsmpi -lws2_32"
+  # Disable MPI I/O on Windows since it causes p4est to crash
+  mpiopts="--enable-mpi --disable-mpiio"
+  # Linker looks for libmsmpi instead of msmpi, copy existing symlink
+  cp -d ${libdir}/msmpi.dll ${libdir}/libmsmpi.dll
```

Comment:
Uhm, the tarball ends up containing msmpi, which looks wrong.

Reply:
I know. However, when we don't do this, the build process works without errors but for some reason no dynamic libraries are created. I've scratched my head as well, but I really don't know how to debug this any further, especially since no errors are thrown. We use the same strategy for building p4est with BB successfully, but if you have another idea, I'm all ears!

Reply:
These are the pertinent messages, I believe, I get when build without linking to `libmsmpi.dll` to `msmpi.sll`:
```
[05:58:43] *** Warning: This system cannot link to static lib archive p4est/src/libp4est.la.
[05:58:43] *** I have the capability to make that library automatically link in when
[05:58:43] *** you link to this library.  But I can only do this if you have a
[05:58:43] *** shared version of the library, which you do not appear to have.
[05:58:43]
[05:58:43] *** Warning: This system cannot link to static lib archive sc/src/libsc.la.
[05:58:43] *** I have the capability to make that library automatically link in when
[05:58:43] *** you link to this library.  But I can only do this if you have a
[05:58:43] *** shared version of the library, which you do not appear to have.
[05:58:43]
[05:58:43] *** Warning: linker path does not have real file for library -lmsmpi.
[05:58:43] *** I have the capability to make that library automatically link in when
[05:58:43] *** you link to this library.  But I can only do this if you have a
[05:58:43] *** shared version of the library, which you do not appear to have
[05:58:43] *** because I did check the linker path looking for a file starting
[05:58:43] *** with libmsmpi and none of the candidates passed a file format test
[05:58:43] *** using a file magic. Last file checked: /workspace/destdir/bin/../../artifacts/70c46814df14fc801299ae69f463aff3f03f9034/bin/libz.dll
[05:58:43] *** The inter-library dependencies that have been dropped here will be
[05:58:43] *** automatically added whenever a program is linked with this library
[05:58:43] *** or is declared to -dlopen it.
[05:58:43]
[05:58:43] *** Since this library must not contain undefined symbols,
[05:58:43] *** because either the platform does not support them or
[05:58:43] *** it was explicitly requested with -no-undefined,
[05:58:43] *** libtool will only create a static version of it.
```

Reply:
I also tried setting
```shell
export LIBS="/workspace/x86_64-w64-mingw32-libgfortran5-cxx11-mpi+microsoftmpi/artifacts/4dec3ecde33187b4e2fcf816ec20e27d18c2c43e/bin/msmpi.dll -lmsmpi -lws2_32"
export LDFLAGS="-L${libdir} -L$WORKSPACE/x86_64-w64-mingw32-libgfortran5-cxx11-mpi+microsoftmpi/artifacts/4dec3ecde33187b4e2fcf816ec20e27d18c2c43e/bin"
```
but to no avail - it still produces the warning above and does not create the shared library :-/

Reply:
> ```sh
> export LIBS="/workspace/x86_64-w64-mingw32-libgfortran5-cxx11-mpi+microsoftmpi/artifacts/4dec3ecde33187b4e2fcf816ec20e27d18c2c43e/bin/msmpi.dll -lmsmpi -lws2_32"
> ```

There are a couple of errors there:

1. `${libdir}/msmpi.dll` is much simpler than hardcoding the full path which points to an artifact directory including its tree hash
2. You either use the absolute path of a shared library ***or*** link it with `-lname`, not both

Reply:
Also,
```console
sandbox:${WORKSPACE} # cat hello.c 
#include <mpi.h>
#include <stdio.h>

int main(int argc, char** argv) {
    // Initialize the MPI environment
    MPI_Init(NULL, NULL);

    // Get the number of processes
    int world_size;
    MPI_Comm_size(MPI_COMM_WORLD, &world_size);

    // Get the rank of the process
    int world_rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);

    // Get the name of the processor
    char processor_name[MPI_MAX_PROCESSOR_NAME];
    int name_len;
    MPI_Get_processor_name(processor_name, &name_len);

    // Print off a hello world message
    printf("Hello world from processor %s, rank %d out of %d processes\n",
           processor_name, world_rank, world_size);

    // Finalize the MPI environment.
    MPI_Finalize();
}
sandbox:${WORKSPACE} # cc -o hello.exe hello.c -lmsmpi -L${libdir}
sandbox:${WORKSPACE} # ${target}-objdump -x hello.exe | grep DLL
 vma:            Hint    Time      Forward  DLL       First
        DLL Name: KERNEL32.dll
        DLL Name: msvcrt.dll
        DLL Name: msmpi.dll
```
The linker here seems to be completely happy with using `-lmsmpi` to link to `msmpi.dll`.

Reply:
> There are a couple of errors there:
> 
>     1. `${libdir}/msmpi.dll` is much simpler than hardcoding the full path which points to an artifact directory including its tree hash
> 
>     2. You either use the absolute path of a shared library _**or**_ link it with `-lname`, not both

Thanks for pointing that out. However, in that case there's another problem, where the autotools-generated `libtool` somehow "drops" the absolute path. For some reasons I don't understand, when called from the Makefile as
```
/bin/sh ./libtool  --tag=CC   --mode=link cc  -O3 -release 2.8.1.5-0b70 -L/workspace/destdir/bin -no-undefined -o src/libsc.la -rpath /workspace/destdir/lib   src/libsc_la-sc.lo src/libsc_la-sc_mpi.lo src/libsc_la-sc_containers.lo src/libsc_la-sc_avl.lo src/libsc_la-sc_string.lo src/libsc_la-sc_unique_counter.lo src/libsc_la-sc_options.lo src/libsc_la-sc_functions.lo src/libsc_la-sc_statistics.lo src/libsc_la-sc_ranges.lo src/libsc_la-sc_io.lo src/libsc_la-sc_amr.lo src/libsc_la-sc_search.lo src/libsc_la-sc_sort.lo src/libsc_la-sc_dmatrix.lo src/libsc_la-sc_blas.lo src/libsc_la-sc_lapack.lo src/libsc_la-sc_bspline.lo src/libsc_la-sc_flops.lo src/libsc_la-sc_random.lo src/libsc_la-sc_getopt.lo src/libsc_la-sc_getopt1.lo src/libsc_la-sc_polynom.lo src/libsc_la-sc_keyvalue.lo src/libsc_la-sc_refcount.lo src/libsc_la-sc_warp.lo src/libsc_la-sc_shmem.lo src/libsc_la-sc_allgather.lo src/libsc_la-sc_reduce.lo src/libsc_la-sc_notify.lo iniparser/src_libsc_la-dictionary.lo iniparser/src_libsc_la-iniparser.lo libb64/src_libsc_la-cencode.lo libb64/src_libsc_la-cdecode.lo  -lz /workspace/destdir/bin/msmpi.dll -lws2_32
```
`libtool` internally calls
```
libtool: link: cc -shared  src/.libs/libsc_la-sc.o src/.libs/libsc_la-sc_mpi.o src/.libs/libsc_la-sc_containers.o src/.libs/libsc_la-sc_avl.o src/.libs/libsc_la-sc_string.o src/.libs/libsc_la-sc_unique_counter.o src/.libs/libsc_la-sc_options.o src/.libs/libsc_la-sc_functions.o src/.libs/libsc_la-sc_statistics.o src/.libs/libsc_la-sc_ranges.o src/.libs/libsc_la-sc_io.o src/.libs/libsc_la-sc_amr.o src/.libs/libsc_la-sc_search.o src/.libs/libsc_la-sc_sort.o src/.libs/libsc_la-sc_dmatrix.o src/.libs/libsc_la-sc_blas.o src/.libs/libsc_la-sc_lapack.o src/.libs/libsc_la-sc_bspline.o src/.libs/libsc_la-sc_flops.o src/.libs/libsc_la-sc_random.o src/.libs/libsc_la-sc_getopt.o src/.libs/libsc_la-sc_getopt1.o src/.libs/libsc_la-sc_polynom.o src/.libs/libsc_la-sc_keyvalue.o src/.libs/libsc_la-sc_refcount.o src/.libs/libsc_la-sc_warp.o src/.libs/libsc_la-sc_shmem.o src/.libs/libsc_la-sc_allgather.o src/.libs/libsc_la-sc_reduce.o src/.libs/libsc_la-sc_notify.o iniparser/.libs/src_libsc_la-dictionary.o iniparser/.libs/src_libsc_la-iniparser.o libb64/.libs/src_libsc_la-cencode.o libb64/.libs/src_libsc_la-cdecode.o   -L/workspace/destdir/bin -lz -lws2_32  -O3   -o src/.libs/libsc-2-8-1-5-0b70.dll -Wl,--enable-auto-image-base -Xlinker --out-implib -Xlinker src/.libs/libsc.dll.a
```

As you can see, `-lz -lws2_32` is passed on to the linker, but the `msmpi.dll` part is dropped.

> The linker here seems to be completely happy with using `-lmsmpi` to link to `msmpi.dll`.

Right. If in the same call as above, I just call the `cc -shared ...` command with `-lmsmpi`, it works and the dll is successfully built 🤷‍♂️

I just pushed another attempt that is not pretty but that works without having to copy `msmpi.dll`.

Reply:
If this doesn't work, I don't think I continue with this, I've already spent an inordinate amount of time debugging the Windows build 😅